src/script/tari_script.rs

```diff
@@ -0,0 +1,345 @@
+    }}
+}
+
+macro_rules! inputs {
+    ($($input:expr),+) => {{
+        use crate::script::ExecutionStack;
+
+        let items = vec![$($input),+];
+        ExecutionStack { items }
+    }}
+}
+
+pub const MAX_STACK_SIZE: usize = 256;
+
+#[derive(Debug, Clone)]
+pub enum StackItem {
+    Number(i64),
+    Commitment(PedersenCommitment),
+    PublicKey(RistrettoPublicKey),
+    Signature(RistrettoSchnorr),
+}
+
+#[derive(Debug, Clone)]
+pub struct ExecutionStack {
+    items: Vec<StackItem>,
+}
+
+impl ExecutionStack {
+    pub fn new() -> Self {
+        ExecutionStack { items: Vec::new() }
+    }
+
+    pub fn size(&self) -> usize {
+        self.items.len()
+    }
+
+    pub fn peek(&self) -> Option<&StackItem> {
+        self.items.last()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.items.is_empty()
+    }
+
+    pub fn pop(&mut self) -> Option<StackItem> {
+        self.items.pop()
+    }
+
+    pub fn push(&mut self, item: StackItem) -> Result<(), ScriptError> {
+        if self.size() >= MAX_STACK_SIZE {
+            return Err(ScriptError::StackOverflow);
+        }
+        self.items.push(item);
+        Ok(())
+    }
+
+    /// Pushes the top stack element down `depth` positions
+    fn push_down(&mut self, depth: usize) -> Result<(), ScriptError> {
+        let n = self.size();
+        if n < depth + 1 {
+            return Err(ScriptError::StackUnderflow);
+        }
+        if depth == 0 {
+            return Ok(());
+        }
+        let top = self.pop().unwrap();
+        self.items.insert(n - depth - 1, top);
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct TariScript {
+    script: Vec<Opcode>,
+}
+
+impl TariScript {
+    pub fn new(script: Vec<Opcode>) -> Self {
+        TariScript { script }
+    }
+
+    pub fn execute(&self, inputs: ExecutionStack) -> Result<(), ScriptError> {
+        // Copy all inputs onto the stack
+        let mut stack = inputs.clone();
+        for opcode in self.script.iter() {
+            self.execute_opcode(opcode, &mut stack)?
+        }
+        TariScript::is_successful(&stack)
+    }
+
+    pub fn as_bytes(&self) -> Vec<u8> {
+        self.script.iter().map(|op| (*op).clone() as u8).collect::<Vec<u8>>()
+    }
+
+    pub fn from_bytes(bytes: &[u8]) -> Result<Self, ScriptError> {
+        let script = bytes
+            .iter()
+            .map(|b| Opcode::from_u8(*b).ok_or(ScriptError::InvalidOpcode))
+            .collect::<Result<Vec<_>, _>>()?;
+        Ok(TariScript { script })
+    }
+
+    /// Calculate the message hash that CHECKSIG uses to verify signatures
+    pub fn script_message(&self, pub_key: &RistrettoPublicKey) -> Result<RistrettoSecretKey, ScriptError> {
+        let b = Blake256::new()
+            .chain(pub_key.as_bytes())
+            .chain(&self.as_bytes())
+            .result()
+            .to_vec();
+        RistrettoSecretKey::from_bytes(&b).map_err(|_| ScriptError::InvalidSignature)
+    }
```

Comment:
90% sure you can save a `to_vec` (clone) here
```suggestion
    pub fn script_message(&self, pub_key: &RistrettoPublicKey) -> Result<RistrettoSecretKey, ScriptError> {
        let b = Blake256::new()
            .chain(pub_key.as_bytes())
            .chain(&self.as_bytes())
            .result();
        RistrettoSecretKey::from_bytes(&b).map_err(|_| ScriptError::InvalidSignature)
    }
```

Reply:
That did not work. I also thought `GenericArray` was referencable as a `&[u8]`

Reply:
`.as_slice` did the trick